### ISSUE TITLE
Feature Request: GPU-Accelerated Gaussian Process Surrogate (via torch / mlr3torch?)

### ISSUE THREAD

**Issue by 15mmerits** [mlr-org/mlr3mbo#218]
**The Problem I Am Facing ;**

Using Bayesian Optimization pipeline for tuning XGBoost/LightGBM models on RTX 2060, AMD 2700X, Linux. While `mlr3mbo` is a phenomenal framework, I am hitting a massive architectural bottleneck when forcing a pure continuous Gaussian Process (GP) surrogate.
Because the legacy spatial packages used for the GP (like `DiceKriging`) rely on standard CPU routines for Cholesky decompositions, they scale at O(N^3). 

As my evaluation rounds increases especially for XGBoost Dart, a massive hidden CPU delay emerges. My XGBoost DART model trains on CUDA in under 30 seconds, But between rounds, the script hangs for 3-4 minutes (vary based on data size and HP search spaces) maxing out CPU cores/threads just to calculate the math for the next point. 

I am aware that `mlr3mbo` bypasses this by falling back to `regr.ranger` for mixed/integer spaces, but I specifically need the mathematical gradient precision of the GP.

**The Solution I'm Asking About ; **

In Python, this O(N^3) GP bottleneck is solved by frameworks like **BoTorch** and **GPyTorch**, which treat the GP matrix algebra natively as CUDA tensors and run the surrogate math on the GPU.

The recent release of `mlr3torch` or any other packages, is there any consideration for building a `torch` backed Gaussian Process surrogate model for `mlr3mbo`? 
Allowing the Bayesian surrogate math to be offloaded to a GPU tensor back-end would completely eliminate the CPU bottleneck for high-dimensional, high-iteration ML tuning in R.

**PS ;**
Right now, the lack of a GPU-accelerated GP back-end seems to be the biggest missing piece for highly scalable Bayesian Optimization (for Gaussian Process) natively in R. If a torch-backed GP is out of scope for mlr3mbo directly, is there another mlr3 extension in the works that might tackle this or I don't know that yet ?

**Comment by 15mmerits** [comment]
```
library(data.table)
library(mlr3)
library(mlr3learners)
library(rBayesianOptimization)

set.seed(2026)
n_rows <- 350000
dt <- as.data.table(matrix(rnorm(n_rows * 60), ncol = 60))
dt$target <- as.factor(sample(c("0", "1"), n_rows, replace = TRUE))

task <- TaskClassif$new(id = "xgb_dart", backend = dt, target = "target", positive = "1")
resampling <- rsmp("holdout", ratio = 0.80)
resampling$instantiate(task)

train_data <- resampling$train_set(1)
valid_data   <- resampling$test_set(1)

# SEARCH SPACE
search_bounds <- list(
  max_depth        = c(4, 10),
  max_bin          = c(128, 512),
  eta_log          = c(log10(0.05), log10(0.2)), # Log scaled
  rate_drop        = c(0.05, 0.30),
  skip_drop        = c(0.0, 0.20),
  subsample        = c(0.7, 1.0),
  colsample_bytree = c(0.6, 1.0),
  lambda_log       = c(-4, log10(5.0)), # Log scaled (1e-4 to 5.0)
  alpha_log        = c(-4, log10(5.0))  # Log scaled (1e-4 to 5.0)
)

# FUNCTION
scoring_fun <- function(max_depth, max_bin, eta_log, rate_drop, skip_drop, 
                        subsample, colsample_bytree, lambda_log, alpha_log) {
  
  iter_start <- Sys.time()
  
  # Learner
  learner <- lrn("classif.xgboost", predict_type = "prob")
  
  learner$param_set$set_values(
    booster = "dart",
    nrounds = 500, 
    early_stopping_rounds = 20,
    
    # 16 Threads + CUDA
    nthread = 16L, # you may try to remove it but its just does not matter at all for CPU overhead
    device = "cuda",
    tree_method = "hist", 
    
    # Fixed Categoricals
    sample_type = "uniform",
    normalize_type = "tree",
    
    # built Search Space
    max_depth = as.integer(round(max_depth)),
    max_bin = as.integer(round(max_bin)),
    eta = 10^eta_log,
    lambda = 10^lambda_log,
    alpha = 10^alpha_log,
    rate_drop = rate_drop,
    skip_drop = skip_drop,
    subsample = subsample,
    colsample_bytree = colsample_bytree
  )
  
  learner$validate <- 0.2
  
  score_auc <- 0.5
  tryCatch({
    learner$train(task, row_ids = train_data)
    preds <- learner$predict(task, row_ids = valid_data)
    score_auc <- preds$score(msr("classif.auc"))
  }, error = function(e) {})
  
  # Timer
  train_time <- as.numeric(difftime(Sys.time(), iter_start, units = "secs"))
  cat(sprintf("XGBoost GPU Training Time: %5.1f sec\n", train_time))
  
  return(list(Score = score_auc, Pred = 0))
}


# GP TUNING (Watch the CPU delay scale after Round 20)

opt_res <- BayesianOptimization(
  FUN = scoring_fun,
  bounds = search_bounds,
  init_points = 20,
  n_iter = 180,
  acq = "ucb",
  kappa = 2.576,
  verbose = TRUE
)
```

**Comment by be-marc** [comment]
You can replace the `DiceKriging` GP with one of the BoTorch learners in [`mlr3extralearners`](https://github.com/mlr-org/mlr3extralearners). They support CUDA. You can also use them as a starting point to implement your own BoTorch learner.

https://github.com/mlr-org/mlr3extralearners/blob/main/R/learner_botorch_regr_singletaskgp.R
https://github.com/mlr-org/mlr3extralearners/blob/main/R/learner_botorch_regr_mixedsingletaskgp.R
https://github.com/mlr-org/mlr3extralearners/blob/main/R/learner_botorch_regr_fullybayesian.R

But very interesting that the `DiceKriging` GP is so slow for you. On how many points is the GP fitted?

**Comment by 15mmerits** [comment]
Thank you for pointing me exactly to those source files 🥇 

I am tuning a 9 to 12 dimensional search space (XGBoost Dart) using sequential loop.
At rounds = 20 evaluations, DiceKriging is lightning fast (virtually instant).
above rounds = 20 to ~100, the math starts taking more than 30 seconds per round.
By rounds = ~100+ evaluations, the O(N^3) matrix inversion creates a severe bottleneck, taking 3 to 4 minutes of 100% threads CPU usage just to calculate the next point via L-BFGS-B optimization I think.
The actual XGBoost DART training on my RTX 2060 only takes less than 25 seconds.

I have a question regarding the BoTorch wrappers;
these rely on reticulate to call PyTorch ?
Will the regr.botorch* learner free the GPU memory b/w surrogate updates so XGBoost doesn't throw an Out-Of-Memory error on a smaller 6GB GPU?

**Comment by be-marc** [comment]
Yes they rely on reticulate. I don't know. Please let us know if you encounter any such problems.
